zkevm-circuits/src/state_circuit/constraint_builder.rs

```diff
@@ -0,0 +1,318 @@
+            cb.build_memory_constraints(q)
+        });
+        self.condition(q.tag_matches(RwTableTag::Stack), |cb| {
+            cb.build_stack_constraints(q)
+        });
+        self.condition(q.tag_matches(RwTableTag::AccountStorage), |cb| {
+            cb.build_account_storage_constraints(q)
+        });
+        self.condition(q.tag_matches(RwTableTag::TxAccessListAccount), |cb| {
+            cb.build_tx_access_list_account_constraints(q)
+        });
+        self.condition(
+            q.tag_matches(RwTableTag::TxAccessListAccountStorage),
+            |cb| cb.build_tx_access_list_account_storage_constraints(q),
+        );
+        self.condition(q.tag_matches(RwTableTag::TxRefund), |cb| {
+            cb.build_tx_refund_constraints(q)
+        });
+        self.condition(q.tag_matches(RwTableTag::Account), |cb| {
+            cb.build_account_constraints(q)
+        });
+        self.condition(q.tag_matches(RwTableTag::AccountDestructed), |cb| {
+            cb.build_account_destructed_constraints(q)
+        });
+        self.condition(q.tag_matches(RwTableTag::CallContext), |cb| {
+            cb.build_call_context_constraints(q)
+        });
+    }
+
+    fn build_general_constraints(&mut self, q: &Queries<F>) {
+        self.require_in_set("tag in RwTableTag range", q.tag(), set::<F, RwTableTag>());
+        self.require_boolean("is_write is boolean", q.is_write());
+    }
+
+    fn build_start_constraints(&mut self, q: &Queries<F>) {
+        self.require_zero("rw_counter is 0 for Start", q.rw_counter.value.clone());
+    }
+
+    fn build_memory_constraints(&mut self, q: &Queries<F>) {
+        self.require_zero("field_tag is 0 for Memory", q.field_tag());
+        self.require_zero("storage_key is 0 for Memory", q.storage_key.encoded.clone());
+        self.require_zero(
+            "read from a fresh key is 0",
+            q.first_access() * q.is_read() * q.value(),
+        );
+        // could do this more efficiently by just asserting address = limb0 + 2^16 *
+        // limb1?
+        for limb in &q.address.limbs[2..] {
+            self.require_zero("memory address fits into 2 limbs", limb.clone());
+        }
+        self.add_lookup(
+            "memory value is a byte",
+            (q.value.clone(), q.lookups.u8.clone()),
+        );
+    }
+
+    fn build_stack_constraints(&mut self, q: &Queries<F>) {
+        self.require_zero("field_tag is 0 for Stack", q.field_tag());
+        self.require_zero("storage_key is 0 for Stack", q.storage_key.encoded.clone());
+        self.require_zero(
+            "first access to new stack address is a write",
+            q.first_access() * q.is_write(),
+        );
+        self.add_lookup(
+            "stack address fits into 10 bits",
+            (q.address.value.clone(), q.lookups.u10.clone()),
+        );
+        self.require_zero(
+            "if call_id doesn't change, stack address change is 0 or 1",
+            q.id_change() * q.address_change(),
```

Comment:
I cannot understand this line.   If call id changes, stack pointer must be same?

Reply:
It turns out I didn't understand it either :D. I have fixed the constraint so that it actually does what it claims.